File: dsa-problems/Tries/solutions/TRI-012-prefix-free-after-inserts.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end = False

class Solution:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, number: str) -> bool:
        node = self.root
        for char in number:
            if node.is_end:
                return False
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        
        if len(node.children) > 0:
            return False
        
        node.is_end = True
        return True

def main():
    import sys
    lines = sys.stdin.read().strip().split('\n')
    n = int(lines[0])
    
    solution = Solution()
    results = []
    for i in range(1, n + 1):
        results.append(solution.insert(lines[i].strip()))
    
    print("[" + ",".join(str(r).lower() for r in results) + "]")

if __name__ == "__main__":
    main()

File: dsa-problems/Tries/solutions/javascript/TRI-012-prefix-free-after-inserts.js
const readline = require('readline');

class TrieNode {
  constructor() {
    this.children = new Map();
    this.isEnd = false;
  }
}

class Solution {
  constructor() {
    this.root = new TrieNode();
  }

  insert(number) {
    let node = this.root;
    for (const char of number) {
      if (node.isEnd) return false;
      if (!node.children.has(char)) {
        node.children.set(char, new TrieNode());
      }
      node = node.children.get(char);
    }
    
    if (node.children.size > 0) return false;
    node.isEnd = true;
    return true;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

const lines = [];
rl.on('line', (line) => {
  lines.push(line);
}).on('close', () => {
  const n = parseInt(lines[0]);
  const solution = new Solution();
  const results = [];
  
  for (let i = 1; i <= n; i++) {
    results.push(solution.insert(lines[i].trim()));
  }
  
  console.log('[' + results.join(',') + ']');
});
